engine/util/util.go

```diff
@@ -331,3 +331,40 @@ func unitStateAction(ctx context.Context, conn *dbus.Conn, unit, action string)
 	}
 	return nil
 }
+
+// edgeCombiner holds the state of the auto edge generator for combining edges.
+type edgeCombiner struct {
+	ae  []engine.AutoEdge
+	ptr int
+}
+
+// Next returns the next automatic edge.
+func (obj *edgeCombiner) Next() []engine.ResUID {
+	if len(obj.ae) <= obj.ptr {
+		panic("shouldn't be called anymore!")
+	}
+	return obj.ae[obj.ptr].Next() // return the next edge
+}
+
+// Test takes the output of the last call to Next() and outputs true if we
+// should continue.
+func (obj *edgeCombiner) Test(input []bool) bool {
+	if len(input) == 0 {
+		return false // no more edges
+	}
+	if !obj.ae[obj.ptr].Test(input) {
```

Comment:
I think this logic is right! But the code could be rewritten to be inlined and nicer please :)